File: threeBody.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Wed Dec 15 17:44:29 2021

@author: tylerpruitt
"""

import matplotlib.pyplot as plt
import numpy as np
import imageio
import os

def eulerStep(x, y, Vx, Vy, Ax, Ay, dt):
    '''Function takes in positions, velocities, accelerations, and dt as inputs
    returns the newly calculated postions and velocities'''
    for i in range(3):
        x[i] += Vx[i] * dt
        y[i] += Vy[i] * dt
        
        Vx[i] += Ax[i] * dt
        Vy[i] += Ay[i] * dt
    
    return x, y, Vx, Vy

def computeAcceleration(masses, x, y):
    '''Function takes in masses and initial positions as inputs
    returns the instantaneous gravitational accelerations of each mass'''
    
    # Initialize the x an y distances between the three objects
    x12 = x[1] - x[0]
    x13 = x[2] - x[0]
    x23 = x[1] - x[0]
    
    # Compute the distances between the objects
    y12 = y[1] - y[0]
    y13 = y[2] - y[0]
    y23 = y[1] - y[0]
    
    r12 = np.sqrt(x12**2 + y12**2)
    r13 = np.sqrt(x13**2 + y13**2)
    r23 = np.sqrt(x23**2 + y23**2)
    
    Ax, Ay = [], []
    
    # Compute the accelerations for objects 1, 2, and 3
    Ax.append(x12 * masses[1] / r12**3 + x13 * masses[2] / r13**3)
    Ay.append(y12 * masses[1] / r12**3 + y13 * masses[2] / r13**3)
    
    Ax.append(-x12 * masses[0] / r12**3 + x23 * masses[2] / r23**3)
    Ay.append(-y12 * masses[0] / r12**3 + y23 * masses[2] / r23**3)
    
    Ax.append(-x13 * masses[0] / r13**3 - x23 * masses[1] / r23**3)
    Ay.append(-y13 * masses[0] / r13**3 - y23 * masses[1] / r23**3)
    
    return Ax, Ay

# Initialize starting conditions (masses, speeds, and positions)
# 0: Sun, 1: Earth, 2: Moon
masses, Vx, Vy, x, y = [], [], [], [], []

print("Enter the masses for the Sun (3.0), Earth (0.01), and Moon(0.0001)")
masses.append(float(input("Sun's mass: ")))
masses.append(float(input("Earth's mass: ")))
masses.append(float(input("Moon's mass: ")))

print("Enter the three inital x velocities (0, 0, 0)")
Vx.append(float(input("Sun's initial x velocity: ")))
Vx.append(float(input("Earth's initial x velocity: ")))
Vx.append(float(input("Moon's initial x velocity: ")))

print("Enter the three inital y velocities (0, 0.5477, 0.6891)")
Vy.append(float(input("Sun's initial y velocity: ")))
Vy.append(float(input("Earth's initial y velocity: ")))
Vy.append(float(input("Moon's initial y velocity: ")))

print("Enter the three inital x positions (0, 10, 10.5)")
x.append(float(input("Sun's initial x position: ")))
x.append(float(input("Earth's initial x position: ")))
x.append(float(input("Moon's initial x position: ")))

print("Enter the three inital y positions (0, 0, 0)")
y.append(float(input("Sun's initial y position: ")))
y.append(float(input("Earth's initial y position: ")))
y.append(float(input("Moon's initial y position: ")))

t = 0
dt = 0.0005
maxTime = 10000

filenames, count = [], 0

impactParameter = 0.1
orbitParameter = 3

print("Loading...")

while np.sqrt((x[1] - x[2])**2 + (y[1] - y[2])**2) <= orbitParameter and t < maxTime and np.sqrt((x[0] - x[1])**2 + (y[0] - y[1])**2) > impactParameter and np.sqrt((x[1] - x[2])**2 + (y[1] - y[2])**2) > impactParameter:
    # Compute accelerations at this time
    Ax, Ay = computeAcceleration(masses, x, y)
    
    # Compute new positions and velocities
    x, y, Vx, Vy = eulerStep(x, y, Vx, Vy, Ax, Ay, dt)
    
    # Loop over all possible quarter of a second times
    for i in range(maxTime * 2 + 1):
        # If t is very close to or equal to quarter of a second time: simulate the dynamics
        if abs(t - 0.25 * i) < 0.001:
            fig = plt.figure()
            plt.plot(x[0], y[0], 'y.', x[1], y[1], 'b.', x[2], y[2], 'w.')
            plt.xlim(-12, 12)
            plt.ylim(-12, 12)
            plt.axis('off')
            fig.set_facecolor('k')
            
            filename = f'{count}.png'
            plt.savefig(filename)
            plt.close()
            
            filenames.append(filename)
            
            count += 1
    
    t += dt

if np.sqrt((x[1] - x[2])**2 + (y[1] - y[2])**2) > 3:
    print("Simulation ended: Moon got too far from Earth")
elif np.sqrt((x[0] - x[1])**2 + (y[0] - y[1])**2) <= impactParameter:
    print("Simulation ended: Earth and Sun collided")
elif np.sqrt((x[1] - x[2])**2 + (y[1] - y[2])**2) <= impactParameter:
    print("Simulation ended: Earth and Moon collided")
else:
    print("Simulation ended: Time ran out")

# Build gif
with imageio.get_writer('orbit.gif', mode='I') as writer:
    for filename in filenames:
        writer.append_data(imageio.imread(filename))
        
# Remove files
for filename in set(filenames):
    os.remove(filename)

File: threeBody.cpp
#include <iostream>
#include <string>
#include <math.h>
#include <vector>
#include <fstream>

using namespace std;

vector<vector<double>> eulerStep(vector<double> x, vector<double> y, vector<double> Vx, vector<double> Vy, vector<double> Ax, vector<double> Ay, double dt) {
    // Function takes in positions, velocities, accelerations, and dt as inputs
    // returns the newly calculated postions and velocities
    for (int i=0;i<3;i++) {
        x[i] += Vx[i] * dt;
        y[i] += Vy[i] * dt;

        Vx[i] += Ax[i] * dt;
        Vy[i] += Ay[i] * dt;
    }

    vector<vector<double>> outputs = {x, y, Vx, Vy};

    return outputs;
}

vector<vector<double>> computeAcceleration(vector<double> masses, vector<double> x, vector<double> y) {
    // Function takes in masses and initial positions as inputs
    // returns the instantaneous gravitational accelerations of each mass

    // Initialize the x and y distances between the three objects
    double x12 = x[1] - x[0];
    double x13 = x[2] - x[0];
    double x23 = x[1] - x[0];

    // Compute the distances between the objects
    double y12 = y[1] - y[0];
    double y13 = y[2] - y[0];
    double y23 = y[1] - y[0];

    double r12 = sqrt(pow(x12, 2) + pow(y12, 2));
    double r13 = sqrt(pow(x13, 2) + pow(y13, 2));
    double r23 = sqrt(pow(x23, 2) + pow(y23, 2));

    vector<double> Ax;
    vector<double> Ay;

    //  Compute the accelearations for object 1
    Ax.push_back(x12 * masses[1] / pow(r12, 3) + x13 * masses[2] / pow(r13, 3));
    Ay.push_back(y12 * masses[1] / pow(r12, 3) + y13 * masses[2] / pow(r13, 3));

    // Compute the accelearations for object 2
    Ax.push_back(-x12 * masses[0] / pow(r12, 3) + x23 * masses[2] / pow(r23, 3));
    Ay.push_back(-y12 * masses[0] / pow(r12, 3) + y23 * masses[2] / pow(r23, 3));

    // Compute the accelearations for object 3
    Ax.push_back(-x13 * masses[0] / pow(r13, 3) - x23 * masses[1] / pow(r23, 3));
    Ay.push_back(-y13 * masses[0] / pow(r13, 3) - y23 * masses[1] / pow(r23, 3));
    
    vector<vector<double>> A = {Ax, Ay};

    return A;
}

void write(vector<double> x, vector<double> y, ofstream &file) {
    vector<string> object = {"Sun", "Earth", "Moon"};
    for (int i=0;i<3;i++) {        
        cout << object[i] << ":" << x[i] << ", " << y[i] << endl;
        
        file << object[i] << ":" << x[i] << ", " << y[i] << endl;
    }
}

int main() {
    // Initialize starting conditions (masses, speeds, and positions)
    // Indices indicate ["Sun"], ["Earth"], ["Moon"]
    vector<double> masses = {3.0, 0.01, 0.0001};
    cout << "Enter the masses for Sun (3.0), Earth (0.01), Moon (0.0001): ";
    cin >> masses[0];
    cin >> masses[1];
    cin >> masses[2];
    
    vector<double> Vx = {0, 0, 0};
    cout << "Enter the 3 initial velocites for Vx (0, 0, 0): ";
    cin >> Vx[0];
    cin >> Vx[1];
    cin >> Vx[2];

    vector<double> Vy = {0, 0.5477, 0.6891};
    cout << "Enter the 3 initial velocites for Vy (0, 0.5477, 0.6891): ";
    cin >> Vy[0];
    cin >> Vy[1];
    cin >> Vy[2];

    vector<double> x = {0, 10, 10.5};
    cout << "Enter the 3 initial positions for x (0, 10, 10.5): ";
    cin >> x[0];
    cin >> x[1];
    cin >> x[2];

    vector<double> y = {0, 0, 0};
    cout << "Enter the 3 initial positions for y (0, 0, 0): ";
    cin >> y[0];
    cin >> y[1];
    cin >> y[2];

    vector<double> Ax;
    vector<double> Ay;

    // Initialize time and time step
    double t = 0;
    double dt = 0.0005;

    ofstream outStream;
    outStream.open("threeBodyProblem.txt");

    while (sqrt( pow(x[1] - x[2], 2)  + pow(y[1] - y[2], 2) ) <= 3 && t < 1000) {
        // Compute the accelerations at this time
        vector<vector<double>> A = computeAcceleration(masses, x, y);
        Ax = A[0];
        Ay = A[1];

        // Compute new positions and velocities
        vector<vector<double>> outputs = eulerStep(x, y, Vx, Vy, Ax, Ay, dt);
        x = outputs[0];
        y = outputs[1];
        Vx = outputs[2];
        Vy = outputs[3];

        // Loop over all possible quarter of a second times
        for (int i=0;i<=2000;i++) {
            // If t is very close to or equal to a quarter of a second time: simulate the dynamics
            if (abs(t - 0.25 * i) < 0.001) {
                write(x, y, outStream);
            }
        }

        t += dt;
    }

    outStream.close();

    return 0;
}

